rfcs/convergence/context-values.md

```diff
@@ -0,0 +1,227 @@
+      <div />
+    </SampleContext.Provider>
+  );
+}
+```
+
+## Problem statement
+
+The problem comes from multiple places:
+
+- (_design issue_) `contextValue` is not a part of `state`, it's useless from component's perspective
+- (_usage issue_) in a case of composition state mutations are not intuitive
+
+The last item is really serious problem as customers should be aware about our implementation details 🕵️‍
+
+```tsx
+// 🛠 In this example I would like to override `size` based on a condition
+function useSuperBazState(props) {
+  const state = useBazState(props);
+
+  // ❌ is broken because `state.contextValue.size` remains the same
+  if (props.reallyBig) {
+    state.size = 'xxl';
+  }
+
+  // ❌ is broken because `state.contextValue` is a stable object and will not trigger
+  //     a rerender
+  if (props.reallyBig) {
+    state.size = 'xxl';
+    state.contextValue.size = 'xxl';
+  }
+
+  // ✅ only this solution will work as expected
+  state.size = 'xxl';
+  state.contextValue = React.useMemo(
+    () => ({
+      ...contextValue,
+      ...(props.reallyBig && { size: 'xxl' }),
+    }),
+    [state.contextValue, props.reallyBig],
+  );
+}
+```
+
+Ideally, we can simply remove `contextValue` from `state` and move it into `render*` function. And it's already done in few components:
+
+```tsx
+function renderBaz(state) {
+  const { open, inline } = state;
+
+  // 💥 this violates rules of hooks: hooks should be called inside components or other hooks
+  const contextValue = React.useMemo(() => ({ inline, open }), [inline, open]);
+
+  return (
+    <SampleContext.Provider value={contextValue}>
+      <div />
+    </SampleContext.Provider>
+  );
+}
+```
+
+The problem is this approach violates [Rules of Hooks](https://reactjs.org/docs/hooks-rules.html) as we are calling `React.useMemo()` inside a function. Actually, it's the reason why `contextValue` is being declared in `use*State()` hooks.
+
+## Solutions
+
+It's clear that we would like to avoid having values for React Context in `state` objects and we don't want to violate Rules of Hooks 🤔
+
+_There is no recomended option yet._
+
+### Option 1: Make `render` functions hooks
+
+We can transform existing render functions to be React hooks:
+
+```diff
+-function renderBaz(state) {
++function useRenderBaz(state) {
+  const { open, inline } = state;
+  // ✅ now we can use hooks
+  const contextValue = React.useMemo(() => ({ inline, open }), [inline, open]);
+
+  // ...
+}
+```
+
+It's one of the easiest options, but it could cause issues in future as it breaks separation of concerns. For example, other hooks can be also called now in `useRender`:
+
+```tsx
+function useRenderBaz(state) {
+  const { open, inline } = state;
+
+  // 💣 other hooks can be also declared there
+  React.useEffect(() => {
+    // ...
+  }, [open]);
+}
+```
+
+- 👎 breaks separation of concerns as allows to use other hooks in render functions
+
+### Option 2: Use `useContextSelector()` everywhere
```

Comment:
I don't think this should be a permanent option, rather just a workaround. There is no sense to use this for all cases